CHAL-40 #done - Add GH Actions workflow to deploy and test staging branch

dvorakjt opened the pull request:
## Checklist
- [x] Include the corresponding Jira issue key and #done in the PR title, like so: "JRA-123 #done Migrate Election Reminders"
- [x] Verify that the code compiles (npm run dev)
- [x] Verify that the project builds (npm run build)
- [x] Verify that all tests pass
- [x] Verify that unit tests cover 100% of the code
- [ ] <s>Create Storybook stories for visual components</s> N/A - no presentation components added
- [ ] <s>Verify that any visual components match the Figma</s> N/A - no presentation components added
- [ ] <s>Test with a screen reader (if applicable)</s> - N/A - no presentation components added
- [ ] <s>Document your code with TSDoc comments</s> - N/A - no TS added
- [ ] <s>Format your code with Prettier</s> - N/A - no TS added

## Overview
The primary purpose of this PR is to add Github Actions workflow files that reset a Supabase database, deploy the application to a preview environment, and then run Selenium tests after accessing that deployment. The workflows run on push to the staging branch, and a separate Supabase project will be created for this branch so that the database can be reset every time this action runs, enabling the end-to-end tests to run against a completely clean environment.

A few other small adjustments were made, such as removing an import of TextEncoder and TextDecoder in the WebCryptoSubtleEncryptor as this class is called from middleware in which Node.js is not available, and adapting the Selenium tests to be able to run in the CI/CD environment (without changing how they can be run locally).

## Test Plan
I set up Supabase and Vercel projects using my personal accounts and verified that the workflow succeeded when I pushed to the staging branch of my fork. I also manually tested the deployed application. As no TypeScript was added, the jest unit tests continue to cover 100% of the codebase.

Screen shot of the passing workflow:
![successful test](https://github.com/user-attachments/assets/eae4f0d1-7a67-4fc5-9bca-2be2d5398c95)

## Follow ups
The appropriate secrets will need to be added to this repo, Supabase and Vercel projects will need to be set up using 8by8 accounts, and then the development branch can be pushed to staging to trigger the workflow.

dvorakjt committed:
added workflow to deploy staging branch

dvorakjt committed:
updated workflow and added TextDecoder to jest environment

dvorakjt committed:
added headless argument to chrome options

dvorakjt committed:
fixed case issues in Selenium tests

dvorakjt committed:
Merge remote-tracking branch 'upstream/development' into gh-actions-for-e2e-tests

ty730 commented on src/__tests__/e2e/homepage_test.py:
```diff
@@ -39,32 +49,27 @@ def test_h1_exists(self):
-        if innerMenu:
+        inner_menu = self.driver.find_element(By.CLASS_NAME, 'styles_inner_container__0JSHj').is_displayed()
+        if inner_menu:
             return None
```
Is there a reason this is returning None? Could this instead check **if not inner_menu** and get rid of the else?

dvorakjt replied:
Yes, this can probably be simplified. Thanks!

ty730 commented on src/__tests__/e2e/homepage_test.py:
```diff
@@ -1,17 +1,21 @@
+"""Runs Selenium tests agains the homepage."""
```
I think you accidentally wrote "agains" instead of "against" :)

dvorakjt replied:
Oops! Yes!

ty730 commented on src/__tests__/e2e/homepage_test.py:
```diff
@@ -83,23 +88,17 @@ def test_content_list(self):
             else:
                 self.fail('TAG_NAME not found!')
```
Is the failure "TAG_NAME not found!" accurate? It seems like this is looking for and element with the class 'styles_hamburger_menu_items__8yFBG'. Should this be something like "Element with class 'styles_hamburger_menu_items__8yFBG' not found"?

dvorakjt replied:
This one slipped past a previous code review, I'll update it. Thanks for pointing this out!

ty730 reviewed (changes requested):
All of your changes in the PR look good. I have some very minor questions about some pre-existing code :)

dvorakjt committed:
fixed issues in homepage_test.py

dvorakjt committed:
fixed typo in method name

dvorakjt committed:
fixed failing tests

dvorakjt committed:
fixed indentation

dvorakjt commented:
Hey Tyler,

I fixed the nits in the Selenium tests and in the process also fixed some Pylint errors. I verified that the refactored tests pass locally and that the workflow still succeeds when the staging branch is pushed to. Let me know if you see anything else!

Thanks!

dvorakjt committed:
simplied hamburger menu lookup

dvorakjt committed:
added comma after last tuple item

ty730 reviewed (approved):
Looks good!

dvorakjt committed:
Merge remote-tracking branch 'upstream/development' into gh-actions-for-e2e-tests
